### DISCUSSION TITLE
(mpfs-rpmsg-bm) Modified make flow using intermediate static library step produces incomplete elf files

### DISCUSSION THREAD

**Discussion by vishakh-mchp** [polarfire-soc/polarfire-soc-documentation#255] [Applications and demos] [answered]
**Introduction**
A reference make flow for the mpfs-rpmsg-bm example is available on Github [here](https://github.com/polarfire-soc/polarfire-soc-amp-examples/tree/main/mpfs-rpmsg-bm). 

The requirement is to modify this make flow to support the use of a static library (_libpolarfire.a_).
The goal is to split the make flow into two steps at the customer site:
STEP 1: Static parts of the project (related to vendor library) will be maintained and updated by a separate team.
STEP 2: Application development team will simply receive an updated static library from team in STEP 1 when the vendor library undergoes updates.   

The modified make flow uses _ar_ command. 
Soft Console is not used  – just the make flow with standard gcc commands.

**Running original make flow**
[The original Github example](https://github.com/polarfire-soc/polarfire-soc-amp-examples/tree/main/mpfs-rpmsg-bm) will create _Remote-Default_ and _Master-Default_ version of elf files.

The remote version is created by default.
Type _make_ in _polarfire-soc-amp-examples/mpfs-rpmsg-bm_ folder:
$ make

For the master version, set the MASTER parameter:
$ make MASTER=MASTER

Results are stored here:
_./Master-Default/mpfs-rpmsg-master.elf
./Remote-Default/mpfs-rpmsg-remote.elf_

**Running modified flow with static library**
The modified make flow setup is described below and all files to reproduce it are uploaded [here](https://microchiptechnology-my.sharepoint.com/:u:/g/personal/vishakh_rayapeta_microchip_com/EbSYqDcpzNFLk-y0eoZ0POUBQCDLlcmzyD_mp4dwgZVrZg?e=ehg0Vu)

STEP 1: Generate libpolarfire.a 

Original Makefile in _polarfire-soc-amp-examples/mpfs-rpmsg-bm_ is modified as below:
![image](https://user-images.githubusercontent.com/71754081/204240899-893d430f-25fa-4dc6-9b13-4908a14c5746.png)

_RISCV_TARGET / TARGET_ is defined in _src/boards/icicle-kit-es/Makefile_.

STEP 2: Compile the demo application using libpolarfire.a
Makefile is inside _src/applications/app_ folder.
Result elf file is inside _src/applications/app/release_
![image](https://user-images.githubusercontent.com/71754081/204240714-e4a89d84-aa17-4bf9-8b86-bd5ad6365403.png)

 
**Comparing the elf files**
I used the _readelf -a_ command to get this information and then grep for ‘FILE’.
Looks like the elf file generated using static library is smaller and missing reference to many C files.

![image](https://user-images.githubusercontent.com/71754081/204240484-4a7d4f5a-a93d-4c46-bc45-1a4dfa9992ec.png)

Why is the elf file generated using the static library incomplete?

**Comment by sanjeevb-bansal** [comment, marked as answer]
Hi Vishakh,

This a generic issue of building and using an archive .a library file , It is not related to  mpfs-rpmsg-bm.

Try  using SoftConsole to include pre compiled archive (.a) file via UI. Good thing in eclipse based SoftConsole, you need not write your own Makefile's etc to use any pre-complied object code.

Hope this helps !!

Thanks
Sanjeev
![image](https://user-images.githubusercontent.com/77011968/205881278-326ec1a5-fb4a-41ea-a734-5f9d853008c9.png)
